pkg/tuf/repo.go

```diff
@@ -0,0 +1,229 @@
+				if err := os.MkdirAll(target, os.ModePerm); err != nil {
+					return err
+				}
+			}
+		// Write out files
+		case tar.TypeReg:
+			fileToWrite, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+			if err != nil {
+				return err
+			}
+			// copy over contents in chunks for security reasons
+			// G110: Potential DoS vulnerability via decompression bomb
+			for {
+				_, err := io.CopyN(fileToWrite, tr, 1024)
+				if err != nil {
+					if errors.Is(err, io.EOF) {
+						break
+					}
+					return err
+				}
+			}
+
+			if err := fileToWrite.Close(); err != nil {
+				return fmt.Errorf("failed to close file %s: %w", target, err)
+			}
+		}
+	}
+	return nil
+}
+
+// From https://github.com/securego/gosec/issues/324
+func sanitizeArchivePath(d, t string) (v string, err error) {
+	v = filepath.Join(d, t)
+	if strings.HasPrefix(v, filepath.Clean(d)) {
+		return v, nil
+	}
+
+	return "", fmt.Errorf("%s: %s", "content filepath is tainted", t)
+}
+
+// UncompressMemFS takes a TUF repository that's been compressed with CompressFS
+// and returns FS backed by memory.
+func UncompressMemFS(src io.Reader) (fs.FS, error) {
+	testFS := fstest.MapFS{}
+
+	zr, err := gzip.NewReader(src)
+	if err != nil {
+		return nil, err
+	}
+	defer zr.Close()
+	tr := tar.NewReader(zr)
+
+	// uncompress each element
+	for {
+		header, err := tr.Next()
+		// EOF is unwrapped
+		//nolint:errorlint
+		if err == io.EOF {
+			break // End of archive
+		}
+		if err != nil {
+			return nil, err
+		}
+		target, err := sanitizeArchivePath("/", header.Name)
+		// validate name against path traversal
+		if err != nil {
+			return nil, err
+		}
+
+		// check the type
+		if header.Typeflag == tar.TypeReg {
+			data := make([]byte, header.Size)
+			_, err := tr.Read(data)
+			// EOF is unwrapped
+			//nolint:errorlint
+			if err != nil && err != io.EOF {
+				return nil, fmt.Errorf("reading file %s : %w", header.Name, err)
+			}
+			testFS[target] = &fstest.MapFile{
+				Data:    data,
+				Mode:    os.FileMode(header.Mode),
+				ModTime: header.ModTime,
+			}
+		}
+	}
+	return testFS, nil
+}
+
+// ClientFromSerializedMirror will construct a TUF client by
+// unserializing the repository and constructing an in-memory client
+// for it.
+func ClientFromSerializedMirror(ctx context.Context, repo []byte, targets string) (*client.Client, error) {
+	r, err := base64.StdEncoding.DecodeString(string(repo))
+	if err != nil {
+		return nil, fmt.Errorf("failed to base64 decode: %w", err)
+	}
+	// unzip/untar the repository.
+	tufFS, err := UncompressMemFS(bytes.NewReader(r))
+	if err != nil {
+		return nil, fmt.Errorf("failed to uncompress: %w", err)
+	}
+	remote, err := client.NewFileRemoteStore(tufFS, targets)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create remote store: %w", err)
+	}
+
+	local := client.MemoryLocalStore()
+	tufClient := client.NewClient(local, remote)
+
+	// TODO(vaikas): What should we do with above tufClient validation
```

Comment:
maybe check the targets

Reply:
Yeah, I'll add that as part of when I add support for bringing those in (next PR or the one after) since this is getting big already 🤣